### ISSUE TITLE
bug(optimization): layer streaming loads the entire checkpoint per layer, per token (#13030)

### ISSUE THREAD

**Issue by mrveiss** [mrveiss/AutoBot-AI#13031]
## Problem

`LayerInferenceEngine` exists to run models larger than available VRAM. Its own module docstring
([layer_inference.py:8-12](autobot-backend/llm_shared/optimization/layer_inference.py#L8-L12)):

> *"During batch/offline inference the entire model need not reside in VRAM simultaneously… Memory
> requirements are therefore bounded by the largest single layer rather than the full model."*

It does the opposite.

**D1 — whole-checkpoint load per layer.**
[layer_inference.py:264](autobot-backend/llm_shared/optimization/layer_inference.py#L264):

```python
full_sd: Dict[str, Any] = torch.load(state_dict_path, map_location="cpu", weights_only=True)
prefix = layer_name + "."
layer_sd = {k[len(prefix):]: v for k, v in full_sd.items() if k.startswith(prefix)}
```

The entire checkpoint is deserialised into RAM, then all but one layer's keys are discarded. Peak
memory is the full model — precisely the thing the module claims to bound.

**D2 — repeated per token.** `_run_layer_loop`
([:477-514](autobot-backend/llm_shared/optimization/layer_inference.py#L477-L514)) is called once per
generated token from the `generate()` loop
([:424-425](autobot-backend/llm_shared/optimization/layer_inference.py#L424-L425)), and calls
`load_layer()` once per layer inside it. A 32-layer model generating 64 tokens deserialises the
checkpoint **2,048 times**. There is no cache of any kind between the loop and the file.

**D5 — the fix is available and unused.**
[:249](autobot-backend/llm_shared/optimization/layer_inference.py#L249) documents support for `.pt`
**or `.safetensors`**, but `torch.load` cannot read safetensors. `grep -rn "safe_open"` across the
backend returns **zero hits**. Safetensors provides lazy per-tensor slicing without full
deserialisation — exactly what D1 needs.

## Impact

High. The module's headline capability is false as written, and the per-token amplification makes it
unusable even where memory is not the constraint. Currently masked because the stack is unregistered
(see the wiring child issue), so there is no production impact yet — but nothing here is fixable by
configuration.

## Fix

1. Replace `torch.load`-whole-checkpoint with `safetensors.safe_open` + per-tensor `get_slice`, so
   only the requested layer's byte ranges are read.
2. Add a bounded slot cache keyed by layer name — fixed slot count, preallocated, **LFU eviction with
   recency tie-break** — reused across the token loop rather than reallocated per layer.
3. Hoist layer acquisition out of the per-token loop; a layer's weights should be read once and reused
   across tokens until evicted.
4. Keep a `torch.load` fallback for `.pt` checkpoints, or narrow the docstring to the formats actually
   supported.

## Acceptance

- Peak RSS during `generate()` is bounded by (largest layer + cache slots + KV), **not** by checkpoint
  size — asserted in a test, not just measured by hand.
- Checkpoint read count is O(layers), not O(layers × tokens).

## Verification required

Evidence in the PR: measured peak RSS before/after on a small real checkpoint, plus the read-count
assertion. Mechanism counts alone (fewer allocations, fewer reads) do not demonstrate the fix —
end-to-end peak memory does.

Part of #13030.

**Comment by mrveiss** [comment]
Closed — verified delivered and present in `Dev_new_gui`.

**Evidence**

| Artifact | Value |
|---|---|
| PR | #13608 (merged) |
| Verified against | `git show origin/Dev_new_gui:<path>` on current base, not the PR diff |

Verified in base — `layer_inference.py:288` uses `torch.load(..., weights_only=True, mmap=True)` and caches into `self._mapped_checkpoints` (declared line 149), with `release_checkpoints()` dropping the cache. Peak memory is bounded by the largest layer plus cache slots rather than the whole checkpoint, and the file is mapped once instead of per layer per token.

---

This issue was fixed but left open because merges into `Dev_new_gui` do not auto-close — `Closes #N` only fires on the default branch. Closed as part of a sweep that cross-referenced merged PRs against open issues and verified each against current base. Every acceptance criterion was checked; partial delivery was left open rather than closed.

**Comment by mrveiss** [comment]
Reopened by the batch-2c closure audit, verified against current `origin/main`. This closed on 2026-08-16 against the old base branch, and its criteria aren't met on main. **D1/D2 are fixed:** `optimization/layer_inference.py:288` loads with `weights_only=True, mmap=True`, tested by `test_checkpoint_is_read_once_across_many_layer_loads`. **Still open:** `.safetensors` can't be read (D5: no `safe_open` anywhere under `llm_shared/optimization/`), and no test asserts peak RSS; the only memory test is `test_get_peak_memory_cpu_returns_zero`.
